multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -0,0 +1,307 @@
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: common.DefaultWorkerCount,
+		}).
+		Complete(r)
+}
+
+// namespaceMapFunc handles Namespace update events (Namespace label change) by enqueuing
+// all Pods in the Namespace into the reconciler processing queue.
+func (r *LabelIdentityReconciler) namespaceMapFunc(ns client.Object) []reconcile.Request {
+	podList := &v1.PodList{}
+	r.Client.List(context.TODO(), podList, client.InNamespace(ns.GetName()))
+	requests := make([]reconcile.Request, len(podList.Items))
+	for i, pod := range podList.Items {
+		requests[i] = reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      pod.GetName(),
+				Namespace: pod.GetNamespace(),
+			},
+		}
+	}
+	return requests
+}
+
+// onPodDelete removes the Pod and label identity mapping from the cache, and
+// updates label identity ResourceExport if necessary (the Pod deletion event
+// causes a label identity no longer present in the cluster).
+func (r *LabelIdentityReconciler) onPodDelete(ctx context.Context, podKey string) error {
+	labelToDelete := r.getLabelToDelete(podKey)
+	if labelToDelete != "" {
+		if err := r.deleteLabelIdentityResourceExport(ctx, labelToDelete); err != nil {
+			return err
+		}
+		r.deleteFromLabelCache(podKey, labelToDelete)
+	}
+	return nil
+}
+
+// onPodCreateOrUpdate updates the Pod and label identity mapping in the cache, and
+// updates label identity ResourceExport if necessary (the Pod creation/update
+// event causes a new label identity to appear in the cluster or a label identity
+// no longer present in the cluster).
+func (r *LabelIdentityReconciler) onPodCreateOrUpdate(ctx context.Context, podKey, normalizedLabel string) error {
+	labelToAdd, labelToDelete := r.getLabelToUpdate(podKey, normalizedLabel)
+	if labelToDelete != "" {
+		if err := r.deleteLabelIdentityResourceExport(ctx, labelToDelete); err != nil {
+			return err
+		}
+		r.deleteFromLabelCache(podKey, labelToDelete)
+	}
+	if labelToAdd != "" {
+		if err := r.addLabelIdentityResourceExport(ctx, labelToAdd); err != nil {
+			return err
+		}
+		r.updateLabelCache(podKey, labelToAdd)
+	}
+	return nil
+}
+
+// getLabelToUpdate gets all label identities to be added and to be deleted due to Pod
+// update event. It is protected with labelMutex so that concurrent Pod update events
+// will not interfere with the label cache update calculations.
+func (r *LabelIdentityReconciler) getLabelToUpdate(podKey, normalizedLabel string) (labelToAdd string, labelToDelete string) {
+	r.labelMutex.RLock()
+	defer r.labelMutex.RUnlock()
+
+	originalLabel, isCached := r.podLabelCache[podKey]
+	if !isCached {
+		return normalizedLabel, ""
+	} else if originalLabel != normalizedLabel {
+		// Pod has updated labels. Check if the original label is stale.
+		if podNames, ok := r.labelToPodsCache[originalLabel]; ok && podNames.Equal(sets.NewString(podKey)) {
```

Comment:
Could we just check podNames size = 1 to be more efficient?

Reply:
I feel this is a tiny overhead but it provides better protection in case the cache goes into some erroneous state for whatever reason. I can refactor it though since by design they should be equivalent.